src/services/pasteEdits.ts

```diff
@@ -0,0 +1,115 @@
+import { addRange } from "../compiler/core";
+import {
+    CancellationToken,
+    SourceFile,
+    Statement,
+    SymbolFlags,
+    TextRange,
+    UserPreferences,
+} from "../compiler/types";
+import { getLineOfLocalPosition } from "../compiler/utilities";
+import {
+    codefix,
+    Debug,
+    fileShouldUseJavaScriptRequire,
+    forEachChild,
+    formatting,
+    getQuotePreference,
+    isIdentifier,
+    textChanges,
+} from "./_namespaces/ts";
+import { addTargetFileImports } from "./refactors/helpers";
+import {
+    addExportsInOldFile,
+    getExistingLocals,
+    getUsageInfo,
+} from "./refactors/moveToFile";
+import {
+    CodeFixContextBase,
+    FileTextChanges,
+    LanguageServiceHost,
+    PasteEdits,
+} from "./types";
+
+const fixId = "providePostPasteEdits";
+/** @internal */
+export function pasteEditsProvider(
+    targetFile: SourceFile,
+    pastedText: string[],
+    pasteLocations: TextRange[],
+    copiedFrom: { file: SourceFile; range: TextRange[]; } | undefined,
+    host: LanguageServiceHost,
+    preferences: UserPreferences,
+    formatContext: formatting.FormatContext,
+    cancellationToken: CancellationToken,
+): PasteEdits {
+    const changes: FileTextChanges[] = textChanges.ChangeTracker.with({ host, formatContext, preferences }, changeTracker => pasteEdits(targetFile, pastedText, pasteLocations, copiedFrom, host, preferences, formatContext, cancellationToken, changeTracker));
+    return { edits: changes, fixId };
+}
+
+function pasteEdits(
+    targetFile: SourceFile,
+    pastedText: string[],
+    pasteLocations: TextRange[],
+    copiedFrom: { file: SourceFile; range: TextRange[]; } | undefined,
+    host: LanguageServiceHost,
+    preferences: UserPreferences,
+    formatContext: formatting.FormatContext,
+    cancellationToken: CancellationToken,
+    changes: textChanges.ChangeTracker,
+) {
+    if (pastedText.length !== 1) {
+        Debug.assert(pastedText.length === pasteLocations.length);
+    }
+    pasteLocations.forEach((paste, i) => {
+        changes.replaceRangeWithText(
+            targetFile,
+            { pos: paste.pos, end: paste.end },
+            pastedText.length === 1 ?
+                pastedText[0] : pastedText[i],
+        );
+    });
+
+    const statements: Statement[] = [];
+
+    let end = targetFile.text.length - 1;
+    let newText = "";
+    for (let i = pasteLocations.length - 1; i >= 0; i--) {
+        newText = pastedText[i] + targetFile.text.slice(pasteLocations[i].end, end) + newText;
+        end = pasteLocations[i].pos;
+    }
+    newText = targetFile.text.slice(0, end) + newText;

```

Comment:
Here's what I meant.

```suggestion
    let newText = targetFile.text;
    for (let i = pasteLocations.length - 1; i >= 0; i--) {
        const { pos, end } = pasteLocations[i];
        newText = newText.slice(0, pos) + pastedText[i] + newText.slice(end);
    }
```